packages/api-server/migrations/20220512033018_refactor_tables.ts

```diff
@@ -0,0 +1,78 @@
+import { Knex } from "knex";
+
+// u32: bigint(i64)
+// u64: decimal(20, 0)
+// u128: decimal(40, 0)
+// u256: decimal(80, 0)
+export async function up(knex: Knex): Promise<void> {
+  await knex.schema
+    .createTable("blocks", function (table: Knex.TableBuilder) {
+      table.decimal("number", null, 0).primary().notNullable();
+      table.binary("hash").notNullable().unique();
+      table.binary("parent_hash").notNullable();
+      table.decimal("gas_limit", null, 0).notNullable();
+      table.decimal("gas_used", null, 0).notNullable();
+      table.binary("miner").notNullable();
+      table.integer("size").notNullable();
+      table.timestamp("timestamp").notNullable();
+    })
+    .createTable("transactions", function (table: Knex.TableBuilder) {
+      table.bigIncrements("id");
+      table.binary("hash").notNullable().unique();
+      table.binary("eth_tx_hash").notNullable().unique();
+      table
+        .decimal("block_number", null, 0)
+        .notNullable()
+        .references("blocks.number");
+      table.binary("block_hash").notNullable();
+      table.integer("transaction_index").notNullable();
+      table.binary("from_address").notNullable().index();
+      table.binary("to_address").index();
+      // value: uint256
+      table.decimal("value", 80, 0).notNullable();
+      table.bigInteger("nonce").notNullable();
+      table.decimal("gas_limit", null, 0);
+      table.decimal("gas_price", null, 0);
+      table.binary("input");
+      table.smallint("v").notNullable();
+      table.binary("r").notNullable();
+      table.binary("s").notNullable();
+      table.decimal("cumulative_gas_used", null, 0);
+      table.decimal("gas_used", null, 0);
+      table.binary("contract_address").index();
+      table.smallint("exit_code").notNullable();
+      table.unique(["block_hash", "transaction_index"], {
+        indexName: "block_hash_transaction_index_idx",
```

Comment:
why use composite index?

Reply:
https://github.com/nervosnetwork/godwoken-web3/blob/compatibility-breaking-changes/crates/indexer/migrations/20210420113516_create_transactions.sql#L31-L32

There are two indexes on `block_hash` before, `(block_hash)` and `(block_hash, transaction_index)`
But when we query transactions only with `block_hash`, it use `(block_hash, transaction_index)` too, so `(block_hash)` index never used before.

1. Preserve `(block_hash, transaction_index)` and delete `(block_hash)`, weather query with `block_hash` or `block_hash` and `transaction_index`, it will queried with indexer.
2. Preserve `(block_hash)` and delete `(block_hash, transaction_index)`, when query with `block_hash` and `transaction_index`, it will query `block_hash` by index and then filter `transaction_index` without index.

For `eth_getTransactionByBlockHashAndIndex` will query with `block_hash` and `transaction_index`, I'm not sure which one is better.

Reply:
One more thing, there is a `unique` constrain on `(block_hash, transaction_index)`, so I preserved it.

Reply:
I see, given that `block_hash` and `transaction_index` are often used together and the unique constraint requirement on those two keys, Preserve `(block_hash, transaction_index)` should be better choice.